### PR TITLE
Implementing functional tests

### PR DESCRIPTION
**Pull request by fidian**
This is pretty big, so I made sure to add documentation explaining the functional tests a bit.

### CODE REVIEW

**Comment by fidian** [comment]
@AbsentSemicolon and @quantumew - bill time to DOT-517 again.

**Comment by fidian** [comment]
BAH!  I have to look into the test failures.  It works on my system and I am certain I can make it work on Travis as well.  Don't review this for now.

File: doc/functional-testing.md [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+### `Promise<response> = startAsync([rel], [title], [options])`
+
+Calls the self discovery endpoint, which should be the basis of all operations with the API.  If a `rel` is specified, then this automatically calls `.follow()`, passing along the link relation, title and options.  The promise is rejected either by a failed self discovery request (probably not going to happen) or by the call to `.follow()`.
+
+    it("calls the self discovery endpoint", () => {
+        return test.startAsync().then((response) => {
+            test.debug(response);
+        });
+    });
```
Comment (quantumew) [inline doc/functional-testing.md]: Awesome documentation.

File: spec/helper/functional-test-async.helper.js [hunk 2]
```diff
@@ -0,0 +1,211 @@
+"use strict";
+
+/**
+ * @typedef {Object} opentoken~functionalTestAsyncResponse
+ * @propety {*} body Can be anything, including objects.
```
Comment (quantumew) [inline spec/helper/functional-test-async.helper.js]: `@property`

Comment (fidian) [inline spec/helper/functional-test-async.helper.js]: Fixed in e8f84d4

File: doc/functional-testing.md [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+
+* `.statusCode` - A number indicating if the response was a success or failure.
+
+* `.uri` - The request URI.
+
+
+Functional Test Instance Methods
+--------------------------------
+
+This uses smoke and mirrors to hide the inner workings of the magic, leaving the audience amazed.
```
Comment (AbsentSemicolon) [inline doc/functional-testing.md]: Cool story bro.

File: doc/functional-testing.md [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+    buff = test.coerseToBuffer({});  // Objects
+    buff = test.coerseToBuffer("string");  // Strings
+    buff = test.coerseToBuffer(buff);  // Buffers
+
+
+### `Promise<response> = test.createAccessCodesAsync()`
+
+Creates a new set of access codes by starting a session (if needed) and making the necessary API calls.  Assigns them to `this.state.accessCode` and `this.state.accessSecret`.
+
+Every response is checked and the promise will be rejected if there's any snags along the way.
```
Comment (AbsentSemicolon) [inline doc/functional-testing.md]: Should be "if there are any snags.." Is is singular, are is plural.

File: doc/functional-testing.md [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+Logs a bunch of information about a response.  Only useful when you are figuring out what's going on with the requests and responses.  All of the output is directed to the console via `console.log` with a tinge of color to make it stand out.
+
+    test.loginAsync().then((response) => {
+        test.debug(response);
+    });
+
+
+### `linkDefinition = test.findLink(links, rel, [title])`
+
+Searches the links and returns a single link definition given the link relation and optionally the link title.  Returns an `Error` object when there's no matching links or too many matching links.
```
Comment (AbsentSemicolon) [inline doc/functional-testing.md]: Should be in the last sentence, "... where there are no matching links..."

File: doc/functional-testing.md [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+### `test.reformatRequestOptions(options)`
+
+Modifies the `options` object so it is in an expected format for the mock request.  Sets several properties and defaults many values.  Assists with URI generation by replacing parameters and appending query string values.
+
+Probably not something that should be called externally.
+
+
+### `Promise<response> = test.requestAsync(options)`
+
+Initiates the fake request.  It does this by first setting up a couple mocks and changing the options that were supplied into a set that the mocks require.
```
Comment (AbsentSemicolon) [inline doc/functional-testing.md]: Don't need the last "that".

File: doc/functional-testing.md [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+
+From here, it makes a promise, fixes a few problems that are created by a conflict between Restify and the mocks, then starts the request.  Events are simulated to send any body on the request and the response events are captured to resolve the promise.  A timer is also started and that timer will help ensure that tests do not run for too long.  The promise is resolved if there is any response and is rejected only if the timer expires.
```
Comment (AbsentSemicolon) [inline doc/functional-testing.md]: Don't need the first "that are"

The next to the last sentence can be rewritten as, "A timer is also started, helping to ensure tests do not run for too long."

File: lib/storage/in-memory.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+            return promise.resolve();
+        }
+
+        return promise.reject(new Error(`No such key: ${filename}`));
+    }
+
+
+    /**
+     * Gets a file from S3 from the configured bucket.
```
Comment (AbsentSemicolon) [inline lib/storage/in-memory.js]: If this is an in memory storage bucket should this say "S3"?

File: lib/storage/in-memory.js [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+
+            return promise.resolve(store[filename]);
+        }
+
+        return promise.reject(new Error(`No such key: ${filename}`));
+    }
+
+
+    /**
+     * Uploads a file to the S3 bucket configured.
```
Comment (AbsentSemicolon) [inline lib/storage/in-memory.js]: Should this say "S3"?

There are other places "S3" is used in comments, should this be that way?

File: spec/helper/functional-test-async.helper.js [hunk 10]
```diff
@@ -0,0 +1,211 @@
+"use strict";
+
+/**
+ * @typedef {Object} opentoken~functionalTestAsyncResponse
+ * @property {*} body Can be anything, including objects.
```
Comment (AbsentSemicolon) [inline spec/helper/functional-test-async.helper.js]: Why does this have a period when other descriptions don't?

File: spec/helper/functional-test-async.helper.js [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+ * so don't go creating new safety layers continually.
+ *
+ * @return {*} Whatever EventEmitter returns
+ */
+function EventEmitterSafetyLayer() {
+    var args;
+
+    args = [].slice.call(arguments);
+
+    // eslint-disable-next-line new-cap
```
Comment (AbsentSemicolon) [inline spec/helper/functional-test-async.helper.js]: Is this the right eslint disable line? You can't calling new here so I'm not sure what you're trying to disable.

File: spec/helper/functional-test.helper.js [hunk 12]
```diff
@@ -0,0 +1,601 @@
...
+                if (options.url.indexOf("?") === -1) {
+                    options.url += "?";
+                } else {
+                    options.url += "&";
+                }
+
+                options.url += `${key}=${options.query[key]}`;
+            });
+        }
+
```
Comment (AbsentSemicolon) [inline spec/helper/functional-test.helper.js]: Too many blank lines.

**Comment by fidian** [comment]
@AbsentSemicolon Fixed.  I completely agree with everything.  The one eslint comment you spotted was a leftover bit of garbage from earlier versions of the same function.

Good job spotting all of these things!

**Comment by AbsentSemicolon** [comment]
Looks good, I approve this pull request.

**Pull request merged**